File: 394-decode-string/decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []

        for i in range(len(s)):

            if s[i] != "]":
                stack.append(s[i])

            else:
                # TAKE THE SUBSTRING 

                st = ""

                while stack and stack[-1] != "[":

                    st = stack.pop() + st
                stack.pop()

                k = ""
                while stack and stack[-1].isdigit():
                    k = stack.pop() + k 
                res = ""

                res = (int(k) * st)
                stack.append(res)
        return "".join(stack)



        

File: 394-decode-string/decode-string.java
class Solution {

    public String decodeString(String s) {

        Stack<String> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {

            char ch = s.charAt(i);

            if (ch != ']') {

                stack.push(String.valueOf(ch));

            } else {

                String st = "";

                while (!stack.isEmpty() && !stack.peek().equals("[")) {
                    st = stack.pop() + st;
                }

                stack.pop();

                String k = "";

                while (!stack.isEmpty() &&
                       Character.isDigit(stack.peek().charAt(0))) {

                    k = stack.pop() + k;
                }

                int repeat = Integer.parseInt(k);

                StringBuilder res = new StringBuilder();

                for (int j = 0; j < repeat; j++) {
                    res.append(st);
                }

                stack.push(res.toString());
            }
        }

        StringBuilder ans = new StringBuilder();

        for (String str : stack) {
            ans.append(str);
        }

        return ans.toString();
    }
}